channels/pkg/cmd/apply_channel.go

```diff
@@ -204,6 +221,80 @@ func getUpdates(ctx context.Context, menu *channels.AddonMenu, k8sClient kuberne
 	return updates, needUpdates, nil
 }
 
+func getDeletions(menu *channels.AddonMenu, channelVersions map[string]*channels.ChannelVersion) map[string]*channels.ChannelVersion {
+	deletions := make(map[string]*channels.ChannelVersion)
+
+	for key, channelVersion := range channelVersions {
+		parts := strings.Split(key, ":")
+		name := parts[1]
+		namespace := parts[0]
+		klog.Infof("Checking for deletion of %q in %q", name, namespace)
+		if addon := menu.FindAddon(name, namespace); addon == nil {
+			deletions[key] = channelVersion
+		}
+
+	}
+
+	return deletions
+}
+
+func buildDeletionPruneSpec(name string) (*api.PruneSpec, error) {
+	spec := &api.PruneSpec{}
+
+	// We add these labels to all objects we manage, so we reuse them for pruning.
+	selectorMap := map[string]string{
+		"app.kubernetes.io/managed-by": "kops",
+		"addon.kops.k8s.io/name":       name,
+	}
+	selector, err := labels.ValidatedSelectorFromSet(selectorMap)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing selector %v: %w", selectorMap, err)
+	}
+
+	// We always include a set of well-known group kinds,
+	// so that we prune even if we end up removing something from the manifest.
+	alwaysPruneGroupKinds := []schema.GroupKind{
+		{Group: "", Kind: "ConfigMap"},
+		{Group: "", Kind: "Service"},
+		{Group: "", Kind: "ServiceAccount"},
+		{Group: "apps", Kind: "Deployment"},
+		{Group: "apps", Kind: "DaemonSet"},
+		{Group: "apps", Kind: "StatefulSet"},
```

Comment:
Aside: it is a pity this one is on here, because it could involve data loss (because it deletes the disks).  That said, I don't think any addons use StatefulSets with data.

Reply:
Doesn't statefullsets preserve disks by default?